src/target/source/interface_rust.cc

```diff
@@ -0,0 +1,380 @@
+      std::string num_elements = NumElements(dtype, size);
+
+      struct_properties.emplace(property.first,
+                                std::make_pair("&mut [" + rust_dtype + "; " + num_elements + "]",
+                                               ".as_ptr() as *mut ::std::os::raw::c_void"));
+    }
+
+    std::vector<std::string> property_names_ordered;
+    for (const String& property_name : property_names) {
+      property_names_ordered.push_back(property_name);
+    }
+    std::reverse(property_names_ordered.begin(), property_names_ordered.end());
+
+    EmitStruct(code_stream, struct_name, struct_properties, property_names_ordered);
+  }
+
+  void EmitDeviceStruct(std::stringstream& code_stream, const std::string& struct_name,
+                        Array<String> devices) {
+    std::unordered_map<std::string, std::pair<std::string, std::string>> struct_properties;
+    std::vector<std::string> property_names_ordered;
+    for (const auto& device : devices) {
+      struct_properties.emplace(device, std::make_pair("*mut ::std::os::raw::c_void", ""));
+      property_names_ordered.push_back(device);
+    }
+    std::reverse(property_names_ordered.begin(), property_names_ordered.end());
+
+    EmitStruct(code_stream, struct_name, struct_properties, property_names_ordered);
+  }
+
+  void EmitIntegerValueConst(std::stringstream& code_stream, const std::string& brief_description,
+                             const std::string& macro_name, int macro_value) {
+    EmitBrief(code_stream, brief_description);
+    std::string macro_name_prefixed = ToRustConstantStyle(macro_name);
+    code_stream << "pub const " << macro_name_prefixed << ": usize = " << macro_value << ";\n";
+  }
+
+  void EmitCRunFunction(std::stringstream& code_stream) {
+    std::string run_function = ToCVariableStyle(PrefixGeneratedName({module_name_, "run"}));
+    code_stream << "extern \"C\" {\n"
+                << "    pub fn " << run_function << "(\n"
+                << "        inputs: *mut Inputs,\n"
+                << "        outputs: *mut Outputs,\n";
+    if (!pools_.empty()) {
+      code_stream << "        workspace_pools: *mut WorkspacePools,\n";
+    }
+    if (!devices_.empty()) {
+      code_stream << "        devices: *mut Devices,\n";
+    }
+    code_stream << "    ) -> i32;\n"
+                << "}\n";
+  }
+
+  void EmitRustRunFunction(std::stringstream& code_stream) {
+    std::string run_function = ToCVariableStyle(PrefixGeneratedName({module_name_, "run"}));
+    code_stream << "/// Entrypoint function for TVM module \"" << module_name_ << "\"\n"
+                << "/// # Arguments\n";
+    if (io_pool_allocations_.empty()) {
+      code_stream << "/// * `inputs` - Input tensors for the module\n";
+      code_stream << "/// * `outputs` - Output tensors for the module\n";
+    }
+
+    if (!pools_.empty()) {
+      code_stream << "/// * `workspace_pools` - Workspace memory pools for the module\n";
+    }
```

Comment:
Given that this is mainly for micro TVM, it would be great to keep it as target/source/microtvm_interface_rust.cc

Reply:
I'd suggest this is out of scope of this PR given it sits parallel with `interface_c.cc`?

Reply:
That would mean that probably we missed that point when reviewing interface_c.cc :)

It would be a great motivator to also scoping that as well

Reply:
I'm not disagreeing with the sentiment, I'm suggesting it's out of scope within this PR as the PR works within the existing framework.

Reply:
This PR proposes a change to generate rust code, and existing framework contains rust code that interacts with tvm in a different way. 

As a result, it would useful to apply proper approaches to clarify the particular file. For example, `target/source/interface_embedded_rust.cc`  would be a fine approach.

Since it clearly signals that there can be other TVM rust interfaces.

Reply:
> As a result, it would useful to apply proper approaches to clarify the intent of the particular file. For example, one possible approach would be target/source/interface_embedded_rust.cc and comments that it intends to generate files that interfaces with embedded rust interface.

This would be different from the existing `interface_c.cc` which is also designed for the embedded use-case, which would mean this patch introduces inconsistency. If we want consistent naming around this we should figure out what we want separately and apply it to both APIs.

Reply:
As the original author of `interface_c.cc` and the Embedded APIs, I will ensure that this is cleaned up properly in line with the other Embedded APIs.